lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php

```diff
@@ -437,12 +486,157 @@ protected static function get_metadata_boolean( $data, $path, $default = false )
 		return $default;
 	}
 
+	/**
+	 * Given a styles array, it extracts the style properties
+	 * and adds them to the $declarations array following the format:
+	 *
+	 * ```php
+	 * array(
+	 *   'name'  => 'property_name',
+	 *   'value' => 'property_value,
+	 * )
+	 * ```
+	 *
+	 * @param array  $styles Styles to process.
+	 * @param array  $settings Theme settings.
+	 * @param array  $properties Properties metadata.
+	 * @param string $selector Selector for styles.
+	 * @return array Returns the modified $declarations.
+	 */
+	protected static function compute_style_properties( $styles, $settings = array(), $properties = null, $selector = null ) {
+		if ( null === $properties ) {
+			$properties = static::PROPERTIES_METADATA;
+		}
+
+		$declarations             = array();
+		$root_variable_duplicates = array();
+
+		if ( empty( $styles ) ) {
+			return $declarations;
+		}
+
+		foreach ( $properties as $css_property => $value_path ) {
+			$value = static::get_property_value( $styles, $value_path );
+
+			if ( strpos( $css_property, '--wp--style--root--' ) === 0 && static::ROOT_BLOCK_SELECTOR !== $selector ) {
+				continue;
+			}
+
+			if ( strpos( $css_property, '--wp--style--root--' ) === 0 ) {
+				$root_variable_duplicates[] = substr( $css_property, strlen( '--wp--style--root--' ) );
+			}
+
+			// Root padding requires special logic to split shorthand values.
+			if ( '--wp--style--root--padding' === $css_property && is_string( $value ) ) {
+
+				$shorthand_top    = '0';
+				$shorthand_right  = '0';
+				$shorthand_bottom = '0';
+				$shorthand_left   = '0';
+		
+				$separate_values = explode( ' ', $value );
```

Comment:
I was thinking about this `explode()` call and the similar one in JS (`styleValue.split( ' ' );`), and the issue https://github.com/WordPress/gutenberg/issues/40132 — would it be worth it to add a parser function so that we can handle values that contain CSS functions like `calc()` `var()`, etc?

Because of nesting it's probably outside the scope of a simple regex, but I hacked together the following in JS to see if it'd be viable:

```js
const inputString = 'var( --my-variable, calc(2px + 3px) ) 2px 3px 0px';

function parseShorthandProps( input ) {
  const values = [];
  
  let openBracket = 0;
  let closedBracket = 0;
  let startIndex = 0;
  let isCollecting = false;
  
  for ( let i = 0; i < input.length; i++ ) {
    if ( ! isCollecting && input[i] !== ' ' ) {
      isCollecting = true;
      startIndex = i;
    }
    
    if ( input[i] === ('(') ) {
      openBracket++;
    }
    
    if ( input[i] === (')') ) {
      closedBracket++;
    }

    if ( isCollecting && ( input[i] === ' ' && openBracket === closedBracket ) || i === input.length - 1 ) {
      isCollecting = false;
      const val = i === input.length -1 ? input.substring( startIndex ) : input.substring( startIndex, i );
      values.push( val );
      openBracket = 0;
      closedBracket = 0;
    }
  }
  
  return values;
}

console.log( parseShorthandProps( inputString ) );
// output: ["var( --my-variable, calc(2px + 3px) )","2px","3px","0px"]
```

If so, perhaps we could add a separate method for that calculation so it doesn't balloon out this function... I was mostly just curious to see if it's possible, but also totally understand if you think it's overkill / too hacky!

Reply:
> would it be worth it to add a parser function so that we can handle values that contain CSS functions like calc() var(), etc?

Nice! This would be helpful right across theme.json, where style input values can be any string.  

Something about it makes me wary, and I'm not sure I can articulate it clearly. I think we'd need to define the patterns that we accept and see how these values function across the application. 

For example, how would `"clamp( var(--my-variable), calc(2px + 3px), 3.4em ) 100px var( --wp-admin-theme-color, 10px ) calc(2px + 2vw)"` affect editor UI controls that expect `value + unit`?

I'm not sure what the right answer is here. Theme authors can add any string they like right now without any guarantee that it'll work. 

I guess the difference is that, if we starting parsing these values out, there might be an expectation that they'll work everywhere.

In the context of padding, we have patterns to define sides `{ top, right, bottom, left }`. 

In the first iteration, would it be safer to assume `string === ''{unit}{value}'` and `array() === [ top, right, bottom, left ]`?  

I don't know. Asking for a friend :trollface:

Reply:
Oooh nice, thanks for writing that! I don't think it's hacky at all.

I'm on the fence as to whether we should add it though; in fact, even the logic already in here might not be necessary if we decide not to support shorthand values at all. There are already issues with them elsewhere (#40132) so it would be good to make a decision and document it explicitly.

Reply:
> Something about it makes me wary, and I'm not sure I can articulate it clearly. I think we'd need to define the patterns that we accept and see how these values function across the application.

Thanks for the thoughts, very nicely articulated!

> I guess the difference is that, if we starting parsing these values out, there might be an expectation that they'll work everywhere.

I think that's definitely the challenge here — if we add a parser, it makes the code more tolerant to values that we don't actually support in any editable way, so I'd think of it as a way to massage values into something we can work with, rather than something we officially support. I don't personally see any issue in saying, "We don't support that yet" and defer adding parsing, etc, to such a point as we want to investigate non `value + unit` values in the property. I was mostly curious to see if we could do it if we need to 😊

Reply:
> There are already issues with them elsewhere (https://github.com/WordPress/gutenberg/issues/40132) so it would be good to make a decision and document it explicitly.

Great point, perhaps it's something to explore in a separate PR. Happy to take your lead on that! I'll copy my code snippet over to that issue for posterity.